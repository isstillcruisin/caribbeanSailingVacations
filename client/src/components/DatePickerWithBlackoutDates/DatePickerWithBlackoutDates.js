import React from "react";
import DayPicker, { DateUtils } from 'react-day-picker';

class DatePickerWithBlackoutDates extends React.Component {
  constructor(props) {
    super(props);
    this.handleDayClick = this.handleDayClick.bind(this);
    this.handleResetClick = this.handleResetClick.bind(this);
    this.state = this.getInitialState();
  }

  getInitialState() {
    return {
      from: undefined,
      to: undefined,
    };
  }

  handleDayClick(day) {
    const range = DateUtils.addDayToRange(day, this.state);
    let blackoutDates = this.props.blackoutDates;
    //Are any of the days in the range overlapping the days in the blackoutDates?
    let found = blackoutDates.find(function(blackoutDate) {
      return blackoutDate.from && blackoutDate.to && 
             range.from && range.to && 
             blackoutDate.from <= range.to && 
             blackoutDate.to >= range.from
    });
    if (!found) {
      this.setState(range);
      this.props.handleSelectedRange(range);
    } else {
      this.setState(Object.assign({}, this.state, { message: 'Unfortunately, those days include some blacked-out dates'}));
    }
  }

  handleResetClick() {
    this.setState(this.getInitialState())
    this.props.handleSelectedRange(this.getInitialState())
  }

  render() {
    const { from, to, message } = this.state;
    const modifiers = { start: from, end: to };
    return (
      <div className="DatePickerWithBlackoutDates">
        <i>{message}</i>
        <p>
          {!from && !to && 'Please select the first day.'}
          {from && !to && 'Please select the last day.'}
          {from &&
            to &&
            `Selected from ${from.toLocaleDateString()} to
                ${to.toLocaleDateString()}`}{' '}
          {from && to && (
            <button className="link" onClick={this.handleResetClick}>
              Reset
            </button>
          )}
        </p>
        <DayPicker
          className="Selectable"
          numberOfMonths={this.props.numberOfMonths || 2}
          selectedDays={[from, { from, to }]}
          modifiers={modifiers}
          onDayClick={this.handleDayClick}
          disabledDays={this.props.blackoutDates}
          month={this.props.month}
        />
      </div>
    )    
  }
}

export default DatePickerWithBlackoutDates;
