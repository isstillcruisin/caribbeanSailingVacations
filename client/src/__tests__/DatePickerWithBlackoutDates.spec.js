import React from "react";
import { create } from "react-test-renderer";
import DatePickerWithBlackoutDates from '../components/DatePickerWithBlackoutDates';

describe("DatePickerWithBlackoutDates component", () => {
  test("Matches the snapshot", () => {
    const alert = create(
      <DatePickerWithBlackoutDates 
        blackoutDates={[
          { from: new Date("December 24, 2019"), to: new Date("January 2, 2020") }
        ]}
        numberOfMonths={1}
      />
    );
    expect(alert.toJSON()).toMatchSnapshot();
  });
});