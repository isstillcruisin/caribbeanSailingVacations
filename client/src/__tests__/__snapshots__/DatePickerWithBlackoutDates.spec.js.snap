// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePickerWithBlackoutDates component Matches the snapshot 1`] = `
<div
  className="DatePickerWithBlackoutDates"
>
  <i />
  <p>
    Please select the first day.
     
  </p>
  <div
    className="DayPicker Selectable"
    lang="en"
  >
    <div
      className="DayPicker-wrapper"
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <div
        className="DayPicker-NavBar"
      >
        <span
          aria-label="Previous Month"
          className="DayPicker-NavButton DayPicker-NavButton--prev"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="0"
        />
        <span
          aria-label="Next Month"
          className="DayPicker-NavButton DayPicker-NavButton--next"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="0"
        />
      </div>
      <div
        className="DayPicker-Months"
      >
        <div
          className="DayPicker-Month"
          role="grid"
        >
          <div
            className="DayPicker-Caption"
            role="heading"
          >
            <div
              onKeyUp={[Function]}
            >
              December 2019
            </div>
          </div>
          <div
            className="DayPicker-Weekdays"
            role="rowgroup"
          >
            <div
              className="DayPicker-WeekdaysRow"
              role="row"
            >
              <div
                className="DayPicker-Weekday"
                role="columnheader"
              >
                <abbr
                  title="Sunday"
                >
                  Su
                </abbr>
              </div>
              <div
                className="DayPicker-Weekday"
                role="columnheader"
              >
                <abbr
                  title="Monday"
                >
                  Mo
                </abbr>
              </div>
              <div
                className="DayPicker-Weekday"
                role="columnheader"
              >
                <abbr
                  title="Tuesday"
                >
                  Tu
                </abbr>
              </div>
              <div
                className="DayPicker-Weekday"
                role="columnheader"
              >
                <abbr
                  title="Wednesday"
                >
                  We
                </abbr>
              </div>
              <div
                className="DayPicker-Weekday"
                role="columnheader"
              >
                <abbr
                  title="Thursday"
                >
                  Th
                </abbr>
              </div>
              <div
                className="DayPicker-Weekday"
                role="columnheader"
              >
                <abbr
                  title="Friday"
                >
                  Fr
                </abbr>
              </div>
              <div
                className="DayPicker-Weekday"
                role="columnheader"
              >
                <abbr
                  title="Saturday"
                >
                  Sa
                </abbr>
              </div>
            </div>
          </div>
          <div
            className="DayPicker-Body"
            role="rowgroup"
          >
            <div
              className="DayPicker-Week"
              role="row"
            >
              <div
                aria-disabled={false}
                aria-label="Sun Dec 01 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={0}
              >
                1
              </div>
              <div
                aria-disabled={false}
                aria-label="Mon Dec 02 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                2
              </div>
              <div
                aria-disabled={false}
                aria-label="Tue Dec 03 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                3
              </div>
              <div
                aria-disabled={false}
                aria-label="Wed Dec 04 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                4
              </div>
              <div
                aria-disabled={false}
                aria-label="Thu Dec 05 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                5
              </div>
              <div
                aria-disabled={false}
                aria-label="Fri Dec 06 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                6
              </div>
              <div
                aria-disabled={false}
                aria-label="Sat Dec 07 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                7
              </div>
            </div>
            <div
              className="DayPicker-Week"
              role="row"
            >
              <div
                aria-disabled={false}
                aria-label="Sun Dec 08 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                8
              </div>
              <div
                aria-disabled={false}
                aria-label="Mon Dec 09 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                9
              </div>
              <div
                aria-disabled={false}
                aria-label="Tue Dec 10 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                10
              </div>
              <div
                aria-disabled={false}
                aria-label="Wed Dec 11 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                11
              </div>
              <div
                aria-disabled={false}
                aria-label="Thu Dec 12 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                12
              </div>
              <div
                aria-disabled={false}
                aria-label="Fri Dec 13 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                13
              </div>
              <div
                aria-disabled={false}
                aria-label="Sat Dec 14 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                14
              </div>
            </div>
            <div
              className="DayPicker-Week"
              role="row"
            >
              <div
                aria-disabled={false}
                aria-label="Sun Dec 15 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                15
              </div>
              <div
                aria-disabled={false}
                aria-label="Mon Dec 16 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                16
              </div>
              <div
                aria-disabled={false}
                aria-label="Tue Dec 17 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                17
              </div>
              <div
                aria-disabled={false}
                aria-label="Wed Dec 18 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                18
              </div>
              <div
                aria-disabled={false}
                aria-label="Thu Dec 19 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                19
              </div>
              <div
                aria-disabled={false}
                aria-label="Fri Dec 20 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                20
              </div>
              <div
                aria-disabled={false}
                aria-label="Sat Dec 21 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                21
              </div>
            </div>
            <div
              className="DayPicker-Week"
              role="row"
            >
              <div
                aria-disabled={false}
                aria-label="Sun Dec 22 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                22
              </div>
              <div
                aria-disabled={false}
                aria-label="Mon Dec 23 2019"
                aria-selected={false}
                className="DayPicker-Day"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                23
              </div>
              <div
                aria-disabled={true}
                aria-label="Tue Dec 24 2019"
                aria-selected={false}
                className="DayPicker-Day DayPicker-Day--disabled"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                24
              </div>
              <div
                aria-disabled={true}
                aria-label="Wed Dec 25 2019"
                aria-selected={false}
                className="DayPicker-Day DayPicker-Day--disabled"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                25
              </div>
              <div
                aria-disabled={true}
                aria-label="Thu Dec 26 2019"
                aria-selected={false}
                className="DayPicker-Day DayPicker-Day--disabled"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                26
              </div>
              <div
                aria-disabled={true}
                aria-label="Fri Dec 27 2019"
                aria-selected={false}
                className="DayPicker-Day DayPicker-Day--disabled"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                27
              </div>
              <div
                aria-disabled={true}
                aria-label="Sat Dec 28 2019"
                aria-selected={false}
                className="DayPicker-Day DayPicker-Day--disabled"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                28
              </div>
            </div>
            <div
              className="DayPicker-Week"
              role="row"
            >
              <div
                aria-disabled={true}
                aria-label="Sun Dec 29 2019"
                aria-selected={false}
                className="DayPicker-Day DayPicker-Day--disabled"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                29
              </div>
              <div
                aria-disabled={true}
                aria-label="Mon Dec 30 2019"
                aria-selected={false}
                className="DayPicker-Day DayPicker-Day--disabled"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                30
              </div>
              <div
                aria-disabled={true}
                aria-label="Tue Dec 31 2019"
                aria-selected={false}
                className="DayPicker-Day DayPicker-Day--disabled"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="gridcell"
                tabIndex={-1}
              >
                31
              </div>
              <div
                aria-disabled={true}
                className="DayPicker-Day DayPicker-Day--disabled DayPicker-Day--outside"
              />
              <div
                aria-disabled={true}
                className="DayPicker-Day DayPicker-Day--disabled DayPicker-Day--outside"
              />
              <div
                aria-disabled={true}
                className="DayPicker-Day DayPicker-Day--outside"
              />
              <div
                aria-disabled={true}
                className="DayPicker-Day DayPicker-Day--outside"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
